### ISSUE TITLE
Order sensitivity in Unions with Recursive TypeVars

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#2145]
**Describe the bug**
```python
from __future__ import annotations

from typing import Mapping, Sequence, TypeVar, Union

S = TypeVar('S')
RecList = Union[Mapping[str, 'RecList[S]'], Sequence['RecList[S]'], S]
RecIntList2 = RecList[int]
RecStrList2 = RecList[str]

T3 = TypeVar('T3', RecIntList2, RecStrList2)


def f3(x: RecIntList2 | RecStrList2) -> None:
    ...


def g3(x: T3):
    return f3(x)


def f4(x: RecStrList2 | RecIntList2) -> None:
    ...


def g4(x: T3):
    return f4(x)
```

The only difference between g4/f4 and g3/f3 is order of union in f4/f3. g3 type checks. g4 has a type error. I tried some simpler cases without recursive types that just used int/str but for those cases they worked fine.

**To Reproduce**
Run pyright on that code.

```
Found 1 source file
/Users/mdrissi/Snapchat/Dev/pyright_bugs/type_var4.py
  /Users/mdrissi/Snapchat/Dev/pyright_bugs/type_var4.py:26:15 - error: Argument of type "T3@g4" cannot be assigned to parameter "x" of type "Mapping[str, RecList] | Sequence[RecList] | str | int" in function "f4"
    Type "T3@g4" cannot be assigned to type "Mapping[str, RecList] | Sequence[RecList] | str | int"
      Type "Mapping[str, RecList]*" cannot be assigned to type "Mapping[str, RecList] | Sequence[RecList] | str | int"
        "Mapping[str, RecList]*" is incompatible with "Sequence[RecList]"
        "Mapping[str, RecList]*" is incompatible with "str"
        "Mapping[str, RecList]*" is incompatible with "int"
          TypeVar "_VT_co@Mapping" is covariant
            Type "Mapping[str, RecList] | Sequence[RecList] | int" cannot be assigned to type "Mapping[str, RecList] | Sequence[RecList] | str"
      Type "Sequence[RecList]*" cannot be assigned to type "Mapping[str, RecList] | Sequence[RecList] | str | int" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
Completed in 0.552sec

```
**Expected behavior**
No errors. Order of union should never matter.

**VS Code extension or command-line**
pyright command line 1.159 on python 3.7.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release. There was a bug in the handling of generic recursive type aliases. I've added a new unit test based on your code sample to prevent regressions now that the bug is fixed.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.160, which I just published. It will also be included in the next release of pylance.
